Enhance <input type=color> with alpha and colorspace=display-p3

annevk opened the pull request:
The new alpha and colorspace attributes give it parity with the 2D canvas API.

As part of this change we do away with the "simple color" concept as that has been replaced by CSS colors now.

Fixes #3400.

---

WebKit can contribute tests as part of implementing this feature.

<s>It would be great to get some help from @svgeesus @tabatkins and @leaverou on resolving #8917 so HTML can clearly link a way to serialize CSS color values (for the Hexadecimal state we can continue to maintain that ourselves). I think we have a plan there, but let me know if there's anything still unclear (prolly best in that issue).</s>

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed):
   * WebKit
   * Chromium
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/48630
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/368059226
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1919718
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=281284
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/35953
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10456/common-microsyntaxes.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/common-microsyntaxes.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/common-microsyntaxes.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )
<a href="https://whatpr.org/html/10456/index.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/index.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/index.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )
<a href="https://whatpr.org/html/10456/indices.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/indices.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/indices.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )
<a href="https://whatpr.org/html/10456/infrastructure.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/infrastructure.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )
<a href="https://whatpr.org/html/10456/input.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/input.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/input.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )
<a href="https://whatpr.org/html/10456/rendering.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">/rendering.html</a>  ( <a href="https://whatpr.org/html/10456/24f5496...298db78/rendering.html" title="Last updated on Oct 16, 2024, 8:14 PM UTC (298db78)">diff</a> )

svgeesus commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+  be manipulated by the end user and does not have to be fully opaque.</p>
+
+  <p>The <dfn element-attr for="input"><code data-x="attr-input-colorspace">colorspace</code></dfn>
+  attribute may be specified to provide user agents with a hint as to what format is desired by the
```
`hint` seems a bit weak here, why is that?

svgeesus commented on source:
```diff
@@ -6762,78 +6762,15 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   </div>
 
 
-  <!--en-GB--><h4 id="colours">Colors</h4>
```
Good to see `simple color` going away

svgeesus commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+   <tbody>
+    <tr>
+     <td><dfn attr-value for="input/colorspace"><code
+     data-x="attr-input-colorspace-hex">hex</code></dfn>
```
So if alpha is also returned by the picker you get two values, one for color and one for alpha, like `#FF047C` and `0.5`? Or is that using 8-digit hex?

svgeesus commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+     <td><dfn attr-value for="input/colorspace"><code
+     data-x="attr-input-colorspace-hex">hex</code></dfn>
+     <td><dfn data-x="attr-input-colorspace-hexadecimal-state">Hexadecimal</dfn>
+     <td>The serialization of the CSS color uses the RGB hexadecimal notation, e.g., "<code
```
```suggestion
     <td>The serialization of the CSS color uses the RGB hexadecimal notation, after conversion to the
     <span>'sRGB'</span> color space, e.g., "<code
```

domenic commented on source:
```diff
@@ -6916,13 +6853,13 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   <hr>
 
   <p class="note">The <span data-x="canvas-context-2d">2D graphics context</span> has a separate
   color syntax that also handles opacity.</p>
```
Double-checking this is still worth including here? Maybe it should be rephrased a bit?

domenic commented on source:
```diff
@@ -140507,6 +140651,11 @@ interface <dfn interface>External</dfn> {
+    <tr>
+     <th> <code data-x="">alpha</code>
+     <td> <code data-x="attr-input-alpha">input</code>
+     <td> Allow the color's alpha component to be picked and manipulated.
```
```suggestion
     <td> Allow the color's alpha component to be picked and manipulated
```

domenic commented on source:
```diff
@@ -140592,6 +140741,12 @@ interface <dfn interface>External</dfn> {
+    <tr>
+     <th> <code data-x="">colorspace</code>
+     <td> <code data-x="attr-input-colorspace">input</code>
+     <td> Provides control over the color space of the web-exposed color.
```
```suggestion
     <td> Provides control over the color space of the web-exposed color
```

I'm not sure what "web-exposed" is doing here, either...

domenic commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+  element's <span data-x="concept-fe-value">value</span> to a string representing the serialization
+  of a CSS color.</p>
+
+  <p class="note" w-nodev>In this state, there is always a CSS color picked, and there is no way to
```
Why add w-nodev here? Seems like useful information for devs.

annevk replied:
This is an existing note that used to be excluded from the dev edition by being inside `<div w-nodev>`. Happy to bring it to the dev edition as part of this overall change though.

domenic commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+  data-x="concept-fe-value">value</span> must be set to the result of <span data-x="serialize a
+  color well control color">serializing a color well control color</span> given the element and the
+  end user's selection. User agents must not allow the end user to set the <span
+  data-x="concept-fe-value">value</span> to the empty string.</p>
```
Several of these sentences seem redundant. Maybe only

> If the user agent provides a user interface for selecting a CSS color, then the value must be set to the result of serializing a color well control color given the element and the end user's selection.

is needed. If you're imaging the user agent allowing users to set the value in some other way, then I guess 

> User agents must not allow the user to set the value to a string that running update a color well control color for the element would not set it to.

could also be included. (Although it's confusing there are kind of two separate algorithms used for determining what is allowed.)

But the empty string one seems definitely redundant.

annevk replied:
I didn't really change anything here from the before version, but I'll remove the empty string variant as I agree that it's redundant.

domenic commented on source:
```diff
@@ -50225,50 +50246,157 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+    in the <span data-x="attr-input-colorspace-hexadecimal-state">Hexadecimal</span> state:
+
+    <ol>
+     <li><p>Set <var>color</var> to <var>color</var> converted to the <span>'srgb'</span> color
```
Can we link to a colorspace conversion algorithm here?

svgeesus replied:
Good idea, I suggest https://drafts.csswg.org/css-color-4/#color-conversion

annevk replied:
This seems reasonable, but we'd also need to change existing 2D canvas text. I'll see about putting up a PR for that as well.

annevk commented:
@domenic observed that for `<input type=color alpha>` we do not have to set the proposed "HTMLCompatible" serialization parameter to true and could instead use the modern serialization. Which would mean instead of `rgba(...)` (that 2D canvas requires as discussed in #8917) you would get `color(srgb ...)` presumably. It would still be limited to 8-bits per color component, but it seems reasonable to me.

domenic commented on source:
```diff
@@ -48152,15 +48139,18 @@ interface <dfn interface>HTMLInputElement</dfn> : <span>HTMLElement</span> {
-  non-negative numbers</span>.</p>
+  <p>The <dfn attribute for="HTMLInputElement"><code
+  data-x="dom-input-colorSpace">colorSpace</code></dfn> IDL attribute must <span>reflect</span> the
+  <code data-x="dom-input-colorspace">colorspace</code> content attribute, <span>limited to only
```
```suggestion
  <code data-x="attr-input-colorspace">colorspace</code> content attribute, <span>limited to only
```

domenic commented on source:
```diff
@@ -50225,50 +50241,164 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+     data-x="attr-input-colorspace-limited-srgb">limited-srgb</code></dfn>
+     <td><dfn data-x="attr-input-colorspace-limited-srgb-state">Limited sRGB</dfn>
+     <td>The CSS color is converted to the <span>'srgb'</span> color space and limited to 8-bits per
+     component, e.g., "<code data-x="">#123456</code>" or "<code data-x="">color(srgb )</code>".
```
This example appears incomplete.

domenic reviewed (commented):
Editorially looking pretty good, although some questions in https://github.com/whatwg/html/issues/3400#issuecomment-2231372016 seem interesting as an uninformed-about-color person.

josepharhar commented:
I think it would take us some time to upgrade the color picker popup in chromium (thanks @lukewarlow for mentioning that DevTools has a better one), but if we can just support serialization etc. in IDLs based on these new attributes in the meantime based on WPTs then that sounds good to me.

lukewarlow commented:
I think it would be unfortunate to support the alpha attribute but not actually allow the user to select an alpha value? But not fully supporting picking display P3 but supporting the serialisation seems like it could be more acceptable?

josepharhar commented:
Yeah I suppose we could easily and quickly add alpha, then hopefully follow up with display p3 later

annevk commented:
I believe I have addressed the questions in #3400 and this is now rebased on top of #8917 so it's using all the latest terms. Marking agenda+ to double check on implementer interest, but if that can be provided out-of-band that would be swell as we've already discussed this a number of times.

svgeesus reviewed (commented):
Looks good

svgeesus commented on source:
```diff
@@ -6883,9 +6820,9 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-    <span data-x="named color">named colors</span>, then return the <span>simple color</span>
-    corresponding to that keyword. <ref>CSSCOLOR</ref></p>
+    <p>If <var>input</var> is an <span>ASCII case-insensitive</span> match for one of the <span
+    data-x="named color">named colors</span>, then return the CSS color corresponding to that
```
Suggest "CSS color" link to https://drafts.csswg.org/css-color-5/#typedef-color

annevk replied:
I don't think that matches how we talk about "CSS color". What we mean here is the parsed representation, the internal object if you will, not the syntax.

svgeesus replied:
Ah okay. Perhaps the &lt;color> type from CSS Typed OM, then?

https://drafts.css-houdini.org/css-typed-om-1/#reify-color

svgeesus commented on source:
```diff
@@ -50268,50 +50284,165 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+    <tr>
+     <td><dfn attr-value for="input/colorspace"><code
+     data-x="attr-input-colorspace-limited-srgb">limited-srgb</code></dfn>
+     <td><dfn data-x="attr-input-colorspace-limited-srgb-state">Limited sRGB</dfn>
```
Is it correct to infer that `limited-srgb` is also limited to (0 .. 1) ie 0 .. 255 range, as well as being limited to 8 bits?

While `display-p3` uses the extended range, so negative values and values greater than 1 (greater than the component value for white) are allowed?

annevk replied:
Yes.

domenic replied:
It might be nice to make the example for display-p3 a couple lines down have a negative component and a component greater than 1, to briefly show this.

svgeesus reviewed (commented):
Looks good overall

domenic commented on source:
```diff
@@ -6799,78 +6799,15 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-  <hr>
+  <!--en-GB--><h4 id="colours">Legacy colors</h4>
 
   <p>Some obsolete legacy attributes parse colors in a more complicated manner, using the
```
Nit: this sentence doesn't make much sense now. ("More complicated than what?")

Also the paragraph has a funny sentence that kind of explains how spec algorithms work, which we can remove. And it declares its parameter in the first line. Maybe modernize it a bit.

There are lots of other old-style things inside the algorithm which you could modernize while you're here, if you wanted.

domenic commented on source:
```diff
@@ -50268,50 +50284,165 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
+  of a CSS color.</p>
+
+  <p class="note">In this state, there is always a CSS color picked, and there is no way to set the
+  value to the empty string.</p>
```
Given all the talk about the empty string later in this section this is a bit confusing. I think it's supposed to mean "no way _for the user_ to set the value"?

josepharhar commented:
I just talked to @mfreed7 @dbaron and @fserb, and we think this is good, so this has chromium implementer interest.

lukewarlow commented:
Having read through the spec one bit still feels slightly ambiguous to me, but this could just be that I'm missing something.

It seems that `<input type="color" alpha>` should serialize to something like `color(srgb 0 0 0 / 1)` but https://whatpr.org/html/10456/input.html#serialize-a-color-well-control-color just calls to https://drafts.csswg.org/css-color/#serializing-color-values I read this to go down the [serialization of sRGB values](https://drafts.csswg.org/css-color/#css-serialization-of-srgb) path, which would result in `rgb(0, 0, 0)`

Is there an obvious reason in the spec that it uses https://drafts.csswg.org/css-color/#serializing-color-function-values instead?
